ResourceLoader: Let resource setup late steps invoke loading in turn

RandomShaper opened the pull request:
The low-level issue in #94586 that the `BinaryMutex` was being tried to be recursively relocked.

Fixes #94586.

RandomShaper committed:
ResourceLoader: Let resource setup late steps invoke loading in turn

AThousandShips reviewed (dismissed):
This makes sense to me though TIWAGOS, ~to me especially the `load_task.cond_var->notify_all();` is a strong indicator that it should be unlocked at that point based on how condition variables work~ (realized this wasn't necessarily relevant either way, as there's nothing wrong with notifying while still locked, or while not locked, just affects some timing)

But my TIWAGOS would mainly apply to if there aren't any critical sections at all in that part, that much I can't speak to

RandomShaper commented:
Thanks to heaven, you smelled something here. I added the lock-unlock too early! I've fixed it. It now happens once the last write has been done, because we don't want other threads to see the state of the task until we're done writing it. It's only afterwards that we enter that last "read-only" segment of code.

RandomShaper commented:
Pushed again since the failed loading branch of the `if` needed it's own handling of the lock. Please re-review because I'm no longer trusting my eyes in this one. 😅

akien-mga reviewed (approved):
Cautiously approving as I don't really understand that code, but know RandomShaper does. More eyes welcome on this, would probably merge soon to include this in 4.3 RC2.

AThousandShips reviewed (approved):
Not familiar with the details in this section but nothing looks obviously weird here, and these parts don't necessarily look like a critical section so should be safe, but TIWAGOS

akien-mga commented:
Thanks!
